fix: Handle Enum naming conflict with built-in types #12761

kapildagur opened the pull request:
### Title

`fix: Handle ENUM and DOMAIN naming conflicts with PostgreSQL built-in types #12761`

### Description

This PR fixes a bug in PostgreSQL where user-defined `ENUM` and `DOMAIN` types with names matching PostgreSQL built-in types (e.g., `text`, `char`) were incorrectly resolved to the built-in type instead of the intended user-defined type.

### Root Cause

PostgreSQL prioritizes `pg_catalog` when resolving unqualified type names. This can shadow user-defined types in other schemas when their names collide with built-in types.

### Solution

* Centralized schema resolution for `ENUM` and `DOMAIN` in `PGIdentifierPreparer.format_type()`.
* The schema resolution logic:

  * Respects explicit schema on the type.
  * Applies schema translation maps when available.
  * Falls back to the dialect’s `default_schema_name` (if configured).
  * Raises `CompileError` if the type name collides with a PostgreSQL built-in type and no schema can be determined.

This avoids collisions with built-in types while maintaining expected behavior for explicitly namespaced and schema-inherited types.

### Test Plan

* New regression tests were added to cover both `ENUM` and `DOMAIN` under different schema configurations, including:

  * Compile-time failure when no schema or default schema is configured.
  * Explicit schema provided.
  * Schema inherited from table definition.
  * Schema inherited from metadata.
  * Schema translation mapping.
* Tests are placed in `test_types` and `test_compiler` and pass across PostgreSQL drivers (`asyncpg`, `pg8000`, `psycopg2`, `psycopg2cffi`).

### Logic Explanation

`PGIdentifierPreparer.format_type()` now performs schema resolution internally instead of requiring explicit `schema` arguments. This ensures consistent handling of:

* Explicit schema and schema translation maps.
* Fallback to dialect’s `default_schema_name`.
* Compile-time errors for unresolved collisions with built-in types.

This centralization avoids duplicate logic across compilers, respects DRY & SRP principles, and ensures correctness in all compilation paths.

Fixes: #12761

KapilDagur commented:
Hii @CaselIT!

I just raised a PR with a test and fix for issue #12761.
Please have a look when you get a chance. Thanks!

zzzeek commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -2775,6 +2785,13 @@ def visit_enum(self, type_, **kw):
+
         return identifier_preparer.format_type(type_)
 
     def visit_DOMAIN(self, type_, identifier_preparer=None, **kw):
```
IIUC we need similar logic for DOMAIN

zzzeek reviewed (changes requested):
Hi !

Thanks for taking this task on.   There's a lot going on with this one that is a lot for a newcomer to take in, so deep breath, there's a bunch of changes I'm requesting here.   The main things are that we need to ensure we dont hardcode a schema name that we cant be sure of, it's better to raise an error, and then after that there's a bunch of existing tests/conventions we need to follow for the tests.  that's mainly it, but I commented everything individually.   I hope you're able to continue work on this, thanks!

KapilDagur commented on test/dialect/postgresql/test_enum_name_confilt_with_internal_types.py:
Test Written inside `test_types.py::NamedTypeTest::test_builtin_name_conflict()`

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision d7390d768cfa47b0d9fbd2d4a6c38c4f2700a1ed of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change d7390d768cfa47b0d9fbd2d4a6c38c4f2700a1ed: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6110

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -2775,11 +2801,39 @@ def visit_enum(self, type_, **kw):
+
+        # Assign default schema for builtin names if schema not already set
+        if (
+            hasattr(type_, "schema")
```
**Michael Bayer** (zzzeek) wrote:

this should just be a type_.schema call.  if the object does not have a .schema attribute at all, that is probably a bug

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6110

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -2775,11 +2801,39 @@ def visit_enum(self, type_, **kw):
+                    f"ENUM with name '{type_.name}' requires an explicit "
+                    "schema when no default_schema_name is configured"
+                )
+            type_.schema = self.dialect.default_schema_name
```
**Michael Bayer** (zzzeek) wrote:

same here, no assignment

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6110

KapilDagur replied:
Same here... object mutation removed, schema is now passed as an argument to `format_type()`

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -2775,11 +2801,39 @@ def visit_enum(self, type_, **kw):
+
+        # Assign default schema for builtin names if schema not already set
+        if (
+            hasattr(type_, "schema")
```
**Michael Bayer** (zzzeek) wrote:

this should just be a type_.schema call. if the object does not have a .schema attribute at all, that is probably a bug

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6110

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3262,9 +3316,15 @@ def __init__(
         super().initialize(connection)
 
+        # Resolve and cache the server current default schema (e.g. "public").
+        self.default_schema_name = self._get_default_schema_name(connection)
```
**Michael Bayer** (zzzeek) wrote:

this is already done by the super().initialize() step (or if not, it's a bug).

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6110

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/named_types.py:
```diff
@@ -497,6 +497,22 @@ def __init__(
+
+        # Only force inherit_schema=False if schema is truly non-None
+        # i.e. we are explicitly binding to a schema to avoid conflicts
+        if kw.get("schema") is not None and (
```
**Michael Bayer** (zzzeek) wrote:

remove all this stuff, we should not have to change how the object is constructed to be able to compile its string representation a certain way.

also there are changes coming down the line for these arguments that would conflict here in any case.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6110

sqla-tester commented on test/dialect/postgresql/test_types.py:
```diff
@@ -467,7 +468,7 @@ def test_schema_inheritance(
 
-        metadata.schema = testing.config.test_schema
+        metadata.schema = testing.config.test_schema_2
         default_schema = testing.config.db.dialect.default_schema_name
```
**Michael Bayer** (zzzeek) wrote:

how come?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6110

sqla-tester commented on test/dialect/postgresql/test_types.py:
```diff
@@ -1240,6 +1241,121 @@ def test_generic_w_some_other_variant(self, metadata, connection):
+        ]
+
+    @testing.variation("type_kind", ["enum", "domain"])
+    def test_builtin_name_conflict(self, connection, type_kind):
```
**Michael Bayer** (zzzeek) wrote:

I know you had multiple tests here before that we removed, but in this case, the "case 1 / 2/ 3/ 4" should likely be separate test methods.  The `@testing.variation` can travel along with them and for now each can have that little bit of setup code at the top (there are ways to make it even more succinct but we can start there).

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6110

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

this is good and the tests are very close. we still need to fix up how the compiler deals with the schema name missing, though, there cant be any mutation of the object inside of compilation.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6110

KapilDagur commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -2383,6 +2383,19 @@ def define_constraint_ondelete_cascade(self, constraint):
+            # attach it to the dialect's default_schema_name.
+            # This mutation is required so that CREATE TYPE places it
+            # in the correct schema and subsequent reflection sees it there.
+            type_.schema = self.dialect.default_schema_name
```
I got your point, Mutating the object inside compilation breaks SRP. I’ve refactored `_get_schema_for_type` and `format_type()` so the schema is now passed as an argument, avoiding object mutation. 
Thanks for the clarification.

KapilDagur commented:
Hey @zzzeek ,

Thanks a lot for patiently reviewing my work and pointing out the needed changes. I made quite a few mistakes (probably more than I’d like to admit 😅), but I’ve fixed them and also tried to clean up a few other things while I was at it.

Really appreciate your guidance and patience!

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 87cae1e7b1a8b7b8a38d48d3d877c3f3a65c0654 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 87cae1e7b1a8b7b8a38d48d3d877c3f3a65c0654 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6110

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

OK.  It looks like everything done in the compiler should be within PGIdentifierPreparer, it already is doing schema resolution, so the enhanced schema resolution stuff should just be there, and this would be a pretty minimal change.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6110

* lib/sqlalchemy/dialects/postgresql/base.py (line 2807): Done
* lib/sqlalchemy/dialects/postgresql/base.py (line 3326): Done
* lib/sqlalchemy/dialects/postgresql/named_types.py (line 508): Done

kapildagur committed:
fix(postgresql): centralize ENUM/DOMAIN schema resolution in PGIdentifierPreparer

Move `_get_schema_for_type` logic out of PGDDLCompiler and PGTypeCompiler,
and consolidate it into `PGIdentifierPreparer._get_schema_for_type`.

`PGIdentifierPreparer.format_type()` now performs schema resolution
internally, removing the need for explicit `schema` arguments.

The logic:
* applies schema translation maps when available
* falls back to the dialect’s `default_schema_name` (skipping "public"
  to preserve legacy behavior)
* raises `CompileError` if a type name conflicts with PostgreSQL builtin
  names and no schema can be determined

This avoids duplicate logic, respects DRY & SRP principles, aligns with
other dialect behaviors, and ensures consistent handling of schema
translation maps, default schema fallback, and builtin type validation.

Tests cover:
* compile-time failure when schema is missing
* explicit schema provided
* schema inheritance via table and metadata

Closes: #12761

kapildagur committed:
test(postgresql): add schema resolution tests for ENUM/DOMAIN with builtin names

Add comprehensive unit tests to verify schema resolution for ENUM and
DOMAIN types when their names conflict with PostgreSQL built-in types.

Covers scenarios:
* raises CompileError when no schema/default is provided
* works with explicit schema specified
* inherits schema from table definition
* inherits schema from metadata

Tests are parameterized via @testing.variation and applied to both ENUM
and DOMAIN. Also includes compiler tests for naming conflicts to ensure
consistent error handling and valid SQL generation.

Closes: #12761

kapildagur committed:
Merge branch 'sqlalchemy:main' into sqlalchemy/fix/pg-internal-datatype-with-enum

kapildagur committed:
revt: f-string reverted as previous one for CREATE & DROP visits

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 63413881132a2e857017dce1f7aca42648699482 of this pull request into gerrit so we can run tests and reviews and stuff

kapildagur committed:
wip:  wiped from  and placed in .

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 793cf27be28a37e8882d89ba6b79570c79f936ac of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 793cf27be28a37e8882d89ba6b79570c79f936ac added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6110

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6110

* lib/sqlalchemy/dialects/postgresql/base.py (line 2816): Done

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6110

* test/dialect/postgresql/test_types.py (line 472): Done

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6110 has been **merged**. Congratulations! :)

zzzeek commented:
> Thanks @zzzeek! I had seen recent patches (Patch 5-6) which directly raising `CompileError` if any conflict enum/domain unqualified name passed without schema passing Is something else planned for future??? This has been my best experience ever, and I would like to continue contributing while learning more about these issues. Thanks for your guidance!!

I dont think so, because i misunderstood this issue originally.  usually when you want to use a name that matches a reserved word in postgresql, you quote it, like `"user"`.   So here I thought if you wanted to name an enum `text`, you'd also quote it, like `"text"` and that would differentiate it from the built in `text` type.

But that's not how PostgreSQL does types, the built in `text` type is a name that comes from a very default namespace `pg_catalog`, whereas our enum named `text` comes from a user-defined default namespace `public`, that does not take precedence over `pg_catalog`.   So given this architecture it makes sense that if someone wants to make a new type that uses the same name as one in the always-default `pg_catalog` namespace, they'd need to schema qualify it.  

The difference between these two methods became apparent once I tried to reflect back an ENUM with the name "text" - it does not come back unless we specify "public" as the schema name, because otherwise, we see the one in `pg_catalog`.   So since an object like `Enum(name="text")` will never come back from `inspect(connection).get_enums()` due to missing schemas, the schema has to be on both sides, `Emum(name="text", schema="public")` and `inspect(connection).get_enums(schema="public")`.

Beyond that I think it's an exceedingly bad idea to use the name of a PostgreSQL built in type as the name of a custom type.
